packages/itwin/grouping-mapping-widget/src/test/GroupingMappingCustomUI.test.tsx

```diff
@@ -0,0 +1,195 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import React from "react";
+import "@testing-library/jest-dom";
+import { render, screen, TestUtils, waitForElementToBeRemoved, within } from "./test-utils";
+import { faker } from "@faker-js/faker";
+import { Groupings } from "../grouping-mapping-widget";
+import type { GroupCollection, Mapping } from "@itwin/insights-client";
+import * as moq from "typemoq";
+import type { IModelConnection, SelectionSet, SelectionSetEvent} from "@itwin/core-frontend";
+import { NoRenderApp } from "@itwin/core-frontend";
+import type { SelectionManager, SelectionScopesManager } from "@itwin/presentation-frontend";
+import { Presentation, SelectionChangeEvent } from "@itwin/presentation-frontend";
+import type { BeEvent } from "@itwin/core-bentley";
+import { setupServer } from "msw/node";
+import { rest } from "msw";
+
+const mockITwinId = faker.datatype.uuid();
+const mockIModelId = faker.datatype.uuid();
+const mockMappingId = faker.datatype.uuid();
+const mockMapping: Mapping = {
+  id: mockMappingId,
+  mappingName: "mOcKmApPiNg1",
+  description: "mOcKmApPiNgDeScRiPtIoN1",
+  createdBy: faker.random.alpha(),
+  createdOn: faker.date.past().toDateString(),
+  modifiedBy: faker.random.alpha(),
+  modifiedOn: faker.date.past().toDateString(),
+  extractionEnabled:false,
+  _links: {
+    imodel: {
+      href: "",
+    },
+  },
+};
+
+const groupsFactory = (): GroupCollection => ({
+  groups: Array.from(
+    { length: faker.datatype.number({ min: 3, max: 5 }) },
+    (_, index) => ({
+      id: `${faker.datatype.uuid()}`,
+      groupName: `mOcKgRoUp${index}`,
+      description: `mOcKgRoUpDeScRiPtIoN${index}`,
+      query: `mOcKgRoUpQuErY${index}`,
+      deleted: false,
+      _links: {
+        imodel: {
+          href: "",
+        },
+        mapping: {
+          href: "",
+        },
+      },
+    })
+  ),
+  _links: {
+    next: undefined,
+    self: {
+      href: "",
+    },
+  },
+});
+
+const connectionMock = moq.Mock.ofType<IModelConnection>();
+const selectionManagerMock = moq.Mock.ofType<SelectionManager>();
+const selectionScopesManagerMock = moq.Mock.ofType<SelectionScopesManager>();
+
+jest.mock("@itwin/appui-react", () => ({
+  ...jest.requireActual("@itwin/appui-react"),
+  useActiveIModelConnection: () => connectionMock.object,
+}));
+
+const server = setupServer();
+
+beforeAll(async () => {
+  // This is required by the i18n module within iTwin.js
+  (global as any).XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest; // eslint-disable-line @typescript-eslint/no-var-requires
```

Comment:
I think the localization module can be replaced with a mock implementation. I think it would be better than this hacky approach.

Reply:
`NoRenderApp.startup()` accepts an object with this property:
```
/** If present, supplies the [[Localization]] for this session. Defaults to [ITwinLocalization]($i18n). */
  localization?: Localization;
```

(https://github.com/iTwin/itwinjs-core/blob/34654961639e0c6d20537c1c411ddf1caf136f7b/core/frontend/src/IModelApp.ts#L109)

Reply:
I used an empty localization. Not sure this will work for other widgets @Guillar1

Reply:
I don't see a reason why it shouldn't work.

Reply:
It should work. When I wrote the tests I didn't realize that you can supply an empty localization through startup.